libs/message-coder/README.md

```diff
@@ -0,0 +1,67 @@
+# `message-coder`
+
+Provides a simple way to encode and decode binary messages from/to JavaScript
+objects.
+
+## Usage
+
+```ts
+import {
+  CoderType,
+  fixedString,
+  literal,
+  message,
+  uint8,
+} from '@votingworks/message-coder';
+
+enum Genre {
+  Blues = 0x00,
+  ClassicRock = 0x01,
+  Country = 0x02,
+  Dance = 0x03,
+  Disco = 0x04,
+  Funk = 0x05,
+  Grunge = 0x06,
+  HipHop = 0x07,
+  Jazz = 0x08,
+  // ...
+}
+
+// define a coder for ID3v1 tags
+const id3v1 = message({
+  header: literal('TAG'),
+  title: fixedString(30),
+  artist: fixedString(30),
+  album: fixedString(30),
+  year: fixedString(4),
+  comment: fixedString(28),
+  zeroByte: literal(0x00),
+  track: uint8(),
+  genre: uint8<Genre>(Genre),
+});
```

Comment:
Is the encoded data dependent on the order of fields in the message object? Is that important?

Reply:
Yes, the order is important. I added a JSDoc comment to `message` that explains it in more detail.

Reply:
Ah ok awesome, that fills in a key gap for my understanding of this whole library. My default mental model treats objects as unordered maps, and I've trained myself to _not_ rely on field order in JS. Do you think I should be updating my mental model or is this a special case?

Reply:
I think this is a special case.

When you later use a `MessageCoder` to _encode_ data, the order of that data does not matter. It conforms to your mental model, and I think one of the main benefits of the abstraction that @eventualbuddha has created here is that you can use the normal mental model when encoding and not worry about order.

USB commands are sent with strictly ordered data, however, and we're just using object notation to represent ordered data. JS allows us to do that because the order in which key-value pairs are defined is respected in browsers and `node`, although looking online and testing in my REPL I see there is an edge cases with numeric keys. I don't think we need to worry about it, but for reference:
```
> const x = { hello: 3, world: 6, 5: 7 }
undefined
> x
{ '5': 7, hello: 3, world: 6 }
```

Reply:
Per [this SO post](https://stackoverflow.com/questions/5525795/does-javascript-guarantee-object-property-order), ES3 did have unordered `Object` properties. In ES2015 that changed to make them be ordered, but with the exception @adghayes pointed out.